# Sistema RAG para Consultas sobre el Modelo 036 de la Agencia Tributaria

## 📋 Descripción del Proyecto

Este proyecto implementa un **sistema de Recuperación Aumentada con Generación (RAG)** especializado en responder preguntas sobre el Modelo 036 de la Agencia Tributaria Española. El sistema combina búsqueda semántica en bases de datos vectoriales con generación de lenguaje natural mediante LLMs.

### 🎯 Objetivo

Desarrollar un asistente conversacional capaz de:
- Responder preguntas específicas sobre el Modelo 036
- Proporcionar información precisa basada en documentación oficial
- Mantener consistencia con las normativas tributarias españolas

---

## 🧠 ¿Qué es RAG (Retrieval-Augmented Generation)?

RAG es una técnica que **mejora las respuestas de los LLMs** combinando dos procesos:

1. **Retrieval (Recuperación)**: Buscar información relevante en una base de conocimientos específica
2. **Generation (Generación)**: Usar esa información como contexto para generar respuestas precisas

### Ventajas de RAG:
- ✅ Respuestas basadas en documentos específicos
- ✅ Reduce alucinaciones del modelo
- ✅ Permite actualizar conocimientos sin reentrenar el modelo
- ✅ Mayor precisión en dominios especializados

---

## 📐 Arquitectura del Sistema

```
┌─────────────────┐
│  Usuario hace   │
│   pregunta      │
└────────┬────────┘
         │
         ▼
┌─────────────────────────────────────────┐
│     1. PREPROCESAMIENTO                 │
│  • Lectura del PDF (Modelo 036)        │
│  • Chunking del documento               │
└────────┬────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────────┐
│     2. VECTORIZACIÓN                    │
│  • Embeddings con Gemini               │
│  • Almacenamiento en ChromaDB          │
└────────┬────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────────┐
│     3. BÚSQUEDA SEMÁNTICA               │
│  • Query del usuario → Vector           │
│  • Búsqueda de similitud en ChromaDB   │
│  • Recuperación de top-k fragmentos    │
└────────┬────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────────┐
│     4. GENERACIÓN CON CONTEXTO          │
│  • LLM recibe: pregunta + contexto     │
│  • Genera respuesta fundamentada       │
└────────┬────────────────────────────────┘
         │
         ▼
┌─────────────────┐
│   Respuesta al  │
│     usuario     │
└─────────────────┘
```

---

## 🔧 Tecnologías Utilizadas

| Componente | Tecnología | Propósito |
|------------|-----------|-----------|
| **LLM** | Google Gemini 2.0 Flash | Generación de respuestas |
| **Embeddings** | Gemini Embedding (exp-03-07) | Vectorización de documentos |
| **Base de Datos Vectorial** | ChromaDB | Almacenamiento y búsqueda de embeddings |
| **Procesamiento PDF** | Unstructured.io | Extracción y chunking de texto |
| **Evaluación** | RAGAS + Gemini | Métricas de precisión de respuestas |


In [ ]:
!pip install openai chromadb google-generativeai "unstructured[pdf]" ragas --quiet

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 2.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 17.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.2/48.2 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.9/18.9 MB 45.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.9/94.9 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 190.9/190.9 kB 8.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 284.2/284.2 kB 12.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 48.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.6/101.6 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
import openai
from langchain_openai import ChatOpenAI
from ragas.dataset_schema import SingleTurnSample
from ragas.llms import LangchainLLMWrapper
from ragas.metrics import AnswerAccuracy

In [ ]:
# Configuración del cliente de OpenAI para usar Gemini
# Aunque estamos usando la librería (SDK) de OpenAI, la redirigimos a la API de Google Gemini
g_api_key = "AIzaSyCgt2QFNzXSMWmKXItpghyigcqie8RECdw"
openai.api_key = g_api_key
openai.base_url = "https://generativelanguage.googleapis.com/v1beta/openai/"

La función **evaluar_respuesta()** usa el modelo Gemini para comparar una respuesta generada con una respuesta de referencia y devuelve un número de precisión entre 0.0 y 1.0 que representa qué tan correcta es la respuesta generada.

Esto permite automatizar la evaluación de tu sistema RAG sin intervención humana.

In [ ]:
# Configura un modelo LLM (en este caso, Gemini 2.0 Flash) para que actúe como evaluador
# Se envuelve con LangchainLLMWrapper para que sea compatible con RAGAS, que es una librería de evaluación de sistemas RAG
# Le dice a RAGAS: usa Gemini como evaluador
evaluator_llm = LangchainLLMWrapper(
    ChatOpenAI(
        model="gemini-2.0-flash", api_key=g_api_key, base_url="https://generativelanguage.googleapis.com/v1beta/openai/"
    )
)

# Se define el evaluador basado en precisión, usando el modelo LLM (Gemini) para comprobar si la response es suficientemente precisa comparada con la reference
scorer = AnswerAccuracy(llm=evaluator_llm)

# Compara la respuesta generada con la de referencia
def evaluar_respuesta(user_input, response, reference):
    """
    Evalúa la precisión de una respuesta utilizando RAGAS.

    Args:
        user_input (str): Pregunta del usuario
        response (str): Respuesta generada por el sistema
        reference (str): Respuesta de referencia (ground truth)

    Returns:
        float: Puntuación de precisión entre 0 y 1
    """
    # Crea una instancia de SingleTurnSample, que es el formato estándar que RAGAS usa para evaluar una sola pregunta-respuesta
    # Crear un sample de evaluación con los datos proporcionados
    sample = SingleTurnSample(user_input=user_input, response=response, reference=reference)

    # Calcular puntuación usando el scorer preexistente, calcula la puntuación de precisión (de 0 a 1) usando el evaluador configurado
    score = scorer.single_turn_score(sample)

    # Devuelve ese score
    return score

La función e**valuar_lista_respuestas()** automatiza la evaluación de todo tu sistema RAG, permitiéndote ver:

Qué tan bien está respondiendo en general.

En qué preguntas específicas acierta o falla (con precisión numérica).

In [ ]:
# Función que se encarga de evaluar las respuestas de los usuarios y calcular la puntuación media.
def evaluar_lista_respuestas(evaluation_list):
    """
    Evalúa una lista de respuestas y calcula la puntuación media.

    Args:
        evaluation_list (list): Lista de diccionarios con las claves:
            - id: Identificador de la respuesta
            - user_input: Pregunta del usuario
            - response: Respuesta generada
            - reference: Respuesta de referencia

    Returns:
        tuple: (puntuación_media, dict_puntuaciones_individuales)
    """
    # Diccionario para almacenar puntuaciones individuales
    # Se inicializa un diccionario vacío para guardar las puntuaciones individuales (id: puntuación)
    # Se inicializa la suma total de puntuaciones para luego calcular el promedio
    puntuaciones = {}
    suma_puntuaciones = 0

    # Se recorre cada entrada (cada pregunta/respuesta) en la lista.
    # Evaluar cada respuesta
    for item in evaluation_list:
        # Obtener puntuación individual
        # Se evalúa la precisión de la respuesta con la función evaluar_respuesta() (que usa RAGAS y el modelo Gemini, función anterior)
        # Se compara la respuesta generada (response) con la esperada (reference)
        score = evaluar_respuesta(
            item['user_input'],
            item.get('response', ''), # Using get() in case response is missing
            item['reference']
        )

        # Almacenar puntuación individual
        puntuaciones[item['id']] = score
        suma_puntuaciones += score

    # Calcular media
    puntuacion_media = suma_puntuaciones / len(evaluation_list) if evaluation_list else 0

    return puntuacion_media, puntuaciones


Diferencias entre las funciones evaluar_respuesta() y evaluar_lista_respuestas()

  **evaluar_respuesta()**
Evalúa una sola respuesta comparándola con la respuesta de referencia.

📤 Devuelve:

*   Un número decimal entre 0 y 1 indicando la precisión (accuracy) de esa única respuesta.

🧠 Útil para:

*   Probar o depurar respuestas individuales.
*   Validar si una respuesta concreta está bien generada.


**evaluar_lista_respuestas()**
Evalúa un conjunto de respuestas (una lista completa) y calcula:

*   La precisión media del conjunto.
*   Las precisiones individuales por cada ítem.

📤 Devuelve:

*   Una tupla con:La precisión media (float)
*   Un diccionario {id: puntuación_individual}.

🧠 Útil para:

*   Evaluar el rendimiento global del sistema RAG.
*   Generar informes de evaluación automáticos.






In [ ]:
# Estos son los datos que se usarán para la evaluación.
evaluation_data = [
    {
        "id": 0,
        "user_input": "¿Quiénes están obligados a presentar la declaración censal de alta (Modelo 036)?",
        "reference": "Están obligados a presentar la declaración de alta los empresarios o profesionales que vayan a comenzar actividades económicas en territorio español, quienes sin ser empresarios/profesionales abonen rentas sujetas a retención o realicen adquisiciones intracomunitarias de bienes sujetas al IVA, los no residentes que operen en España mediante establecimiento permanente o satisfagan rentas sujetas a retención, las entidades en régimen de atribución de rentas extranjeras con presencia en España, los socios/herederos/comuneros/partícipes de entidades en atribución de rentas con obligaciones tributarias derivadas, y los no establecidos en territorio de aplicación del IVA que sean sujetos pasivos del mismo (salvo exoneración).",
    },
    {
        "id": 1,
        "user_input": "¿Para qué sirve la declaración censal de modificación?",
        "reference": "La declaración de modificación se debe presentar cuando varíe cualquiera de los datos identificativos del declarante (sirviendo para comunicar el cambio de domicilio fiscal) o cualquier otro dato consignado en la declaración de alta o en una modificación posterior.",
    },
    {
        "id": 2,
        "user_input": "¿En qué plazo general se debe presentar la declaración de modificación?",
        "reference": "El plazo general para presentar la declaración de modificación es de un mes, a contar desde el día siguiente a aquél en que se hayan producido los hechos que determinan su presentación.",
    },
    {
        "id": 3,
        "user_input": "¿Dónde se puede presentar el modelo 036 en formato impreso?",
        "reference": "La presentación en impreso del modelo 036 podrá realizarse mediante entrega directa en la oficina correspondiente (Administración o Delegación de la Agencia Tributaria, Unidad Regional de Gestión de Grandes Empresas o Delegación Central de Grandes Contribuyentes según corresponda) o enviándolo por correo certificado.",
    },
    {
        "id": 4,
        "user_input": "¿Cuándo es obligatorio presentar el modelo 036 por vía telemática?",
        "reference": "La presentación del modelo 036 debe realizarse de forma obligatoria por vía telemática a través de Internet en el supuesto de solicitud de inscripción o baja en el Registro de devolución mensual del Impuesto sobre el Valor Añadido.",
    },
    {
        "id": 5,
        "user_input": "¿Qué plazo tienen los herederos para comunicar la modificación de titularidad de derechos y obligaciones tras el fallecimiento del obligado tributario?",
        "reference": "Los herederos deberán comunicar la modificación de la titularidad de los derechos y obligaciones con trascendencia tributaria vigentes con terceros en el plazo de seis meses desde el fallecimiento del obligado tributario.",
    },
]

In [ ]:
# Respuestas de ejemplo como las que vosotros deberéis generar usando un LLM
respuestas_de_ejemplo = [
    "Tienen que presentar el modelo 036 todos los que quieran hacer negocios en España, así como algunos ciudadanos que ganen dinero fuera pero quieran declarar aquí, incluso si no tienen empresa. También se menciona algo de los que pagan IVA, aunque no siempre es necesario. Las asociaciones internacionales pueden que también tengan que hacerlo, dependiendo del caso.",
    "La declaración de modificación se presenta para cambiar cosas como el nombre o el teléfono en Hacienda. También sirve si te mudas, aunque no es obligatorio en todos los casos. A veces se puede usar para modificar datos bancarios, aunque eso depende del tipo de actividad que tengas.",
    "En términos generales, se debe entregar el cambio de información en menos de un mes desde que pasó algo importante, como un cambio de domicilio o si se empieza una nueva actividad secundaria. Aunque si el cambio es menor, algunas personas lo presentan más tarde sin consecuencias.",
    "El modelo 036 en papel se puede llevar a correos o a la oficina más cercana de Hacienda, siempre que no se tenga certificado digital. A veces también se permite entregarlo en el Ayuntamiento, aunque eso solo pasa en municipios pequeños o si hay convenios especiales.",
    "Solo es obligatorio presentar el 036 online si tienes que hacer devoluciones de IVA de manera mensual. En otros casos puedes decidir entre Internet o papel. Pero si eres autónomo, normalmente te lo exigen online por defecto, aunque hay excepciones si no tienes acceso a Internet.",
    "Tras un fallecimiento, los herederos tienen hasta seis meses para hacer gestiones relacionadas con contratos, impuestos y otras obligaciones. Sin embargo, si no hay testamento, puede que se necesite una prórroga. Además, este plazo se puede suspender si se inicia un proceso judicial por la herencia.",
]


In [ ]:
# Ejemplo del formato esperado para evaluar.
lista_ejemplo_evaluacion = [elem | {"response": respuestas_de_ejemplo[i]} for i, elem in enumerate(evaluation_data)]
lista_ejemplo_evaluacion

[{'id': 0,
  'user_input': '¿Quiénes están obligados a presentar la declaración censal de alta (Modelo 036)?',
  'reference': 'Están obligados a presentar la declaración de alta los empresarios o profesionales que vayan a comenzar actividades económicas en territorio español, quienes sin ser empresarios/profesionales abonen rentas sujetas a retención o realicen adquisiciones intracomunitarias de bienes sujetas al IVA, los no residentes que operen en España mediante establecimiento permanente o satisfagan rentas sujetas a retención, las entidades en régimen de atribución de rentas extranjeras con presencia en España, los socios/herederos/comuneros/partícipes de entidades en atribución de rentas con obligaciones tributarias derivadas, y los no establecidos en territorio de aplicación del IVA que sean sujetos pasivos del mismo (salvo exoneración).',
  'response': 'Tienen que presentar el modelo 036 todos los que quieran hacer negocios en España, así como algunos ciudadanos que ganen diner

In [ ]:
# Ejempo de evaluación
resultado_evaluacion = evaluar_lista_respuestas(lista_ejemplo_evaluacion)
print(f"Precisión media de las respuestas: {resultado_evaluacion[0]} \n\n Resultados individuales: \n\n {resultado_evaluacion[1]}")

Precisión media de las respuestas: 0.625 

 Resultados individuales: 

 {0: 0.5, 1: 0.5, 2: 0.5, 3: 1.0, 4: 0.5, 5: 0.75}


In [ ]:
from langchain.embeddings import OpenAIEmbeddings
from langchain.vectorstores import FAISS
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.document_loaders import UnstructuredPDFLoader
from langchain.chains import RetrievalQA
from langchain_openai import ChatOpenAI

# 1. Leer y trocear un documento PDF (Modelo 036) usando unstructured.

**Extracción y Chunking del Documento**

El procesamiento inteligente del PDF es crítico para el rendimiento del sistema. Los LLMs tienen límites de tokens en su ventana de contexto, y los embeddings funcionan mejor con textos semánticamente coherentes. Una estrategia de chunking efectiva mejora significativamente la precisión de la búsqueda.

In [ ]:
# Importamos la biblioteca unstructured para procesar PDF y convertirlos a texto
# partition es una función que extrae contenido de varios tipos de documentos
from unstructured.partition.auto import partition
filename = "./sample_data/instr_mod036.pdf"
from unstructured.chunking.title import chunk_by_title


In [ ]:
# Procesamos el PDF y lo convertimos en elementos de texto
# strategy="auto" indica que unstructured debe determinar automáticamente la mejor forma de procesar
modelo_36_docs = partition(filename=filename, content_type="application/pdf", strategy="auto", languages=["es"])

In [ ]:
# Visualizamos los elementos extraídos del PDF
# Cada elemento representa un fragmento de texto del documento (párrafo, título, etc.)
modelo_36_docs

In [ ]:
# Metadatos extraídos automáticamente por unstructured
modelo_36_docs[0].metadata.__dict__.keys()

dict_keys(['coordinates', 'file_directory', 'filename', 'languages', 'last_modified', 'links', 'page_number', '_known_field_names', 'filetype'])

In [ ]:
# Mostramos una muestra del contenido extraído para verificar la calidad
print("\n\n".join([f"Page number: {doc.metadata.page_number} || "+str(doc) for doc in modelo_36_docs]))

Page number: 1 || INSTRUCCIONES PARA CUMPLIMENTAR LA DECLARACIÓN CENSAL MODELO 036.

Page number: 1 || Identifique todas las páginas en las que haya cumplimentado alguna casilla con el NIF (si dispone de él) y el nombre y apellidos o razón o denominación social del titular de la declaración.

Page number: 1 || OBLIGADOS A DECLARAR

Page number: 1 || Declaración de alta:

Page number: 1 || Empresarios o profesionales que vayan a comenzar el ejercicio de una o varias

Page number: 1 || actividades económicas en territorio español.

Page number: 1 || Quienes, no actuando como empresarios o profesionales, abonen rentas sujetas a retención o ingreso a cuenta o realicen adquisiciones intracomunitarias de bienes sujetas al IVA.

Page number: 1 || Los no residentes que operen en territorio español mediante establecimiento permanente o satisfagan en dicho territorio rentas sujetas a retención o ingreso a cuenta, así como las entidades en régimen de atribución de rentas constituidas en el extran

In [ ]:
# Aplica chunking por títulos para dividir mejor los contenidos semánticos
#chunked_docs = chunk_by_title(modelo_36_docs)

# Extraer los textos ya troceados
#modelo_36_raw_docs = [doc.text for doc in chunked_docs if doc.text.strip() and doc.text != ""]
modelo_36_raw_docs = [doc.text for doc in modelo_36_docs if doc.text != ""]

In [ ]:
# Visualizamos los textos ya troceados
modelo_36_raw_docs

['INSTRUCCIONES PARA CUMPLIMENTAR LA DECLARACIÓN CENSAL MODELO 036.',
 'Identifique todas las páginas en las que haya cumplimentado alguna casilla con el NIF (si dispone de él) y el nombre y apellidos o razón o denominación social del titular de la declaración.',
 'OBLIGADOS A DECLARAR',
 'Declaración de alta:',
 'Empresarios o profesionales que vayan a comenzar el ejercicio de una o varias',
 'actividades económicas en territorio español.',
 'Quienes, no actuando como empresarios o profesionales, abonen rentas sujetas a retención o ingreso a cuenta o realicen adquisiciones intracomunitarias de bienes sujetas al IVA.',
 'Los no residentes que operen en territorio español mediante establecimiento permanente o satisfagan en dicho territorio rentas sujetas a retención o ingreso a cuenta, así como las entidades en régimen de atribución de rentas constituidas en el extranjero con presencia en territorio español. Asimismo, los establecimientos permanentes en territorio español de las persona

In [ ]:
# Mostramos una muestra del contenido extraído para verificar la calidad
print("\n\n".join([f"Page number: {doc.metadata.page_number} || "+str(doc) for doc in chunked_docs if doc.text.strip()]))

Page number: 1 || INSTRUCCIONES PARA CUMPLIMENTAR LA DECLARACIÓN CENSAL MODELO 036.

Identifique todas las páginas en las que haya cumplimentado alguna casilla con el NIF (si dispone de él) y el nombre y apellidos o razón o denominación social del titular de la declaración.

OBLIGADOS A DECLARAR

Page number: 1 || Declaración de alta:

Empresarios o profesionales que vayan a comenzar el ejercicio de una o varias

actividades económicas en territorio español.

Quienes, no actuando como empresarios o profesionales, abonen rentas sujetas a retención o ingreso a cuenta o realicen adquisiciones intracomunitarias de bienes sujetas al IVA.

Page number: 1 || Los no residentes que operen en territorio español mediante establecimiento permanente o satisfagan en dicho territorio rentas sujetas a retención o ingreso a cuenta, así como las entidades en régimen de atribución de rentas constituidas en el extranjero con presencia en territorio español. Asimismo, los establecimientos permanentes en te

## **2. Arquitectura de Base de Datos Vectorial**

Los embeddings son representaciones vectoriales que capturan el significado semántico del texto, permitiendo búsquedas basadas en similitud conceptual en lugar de coincidencias de palabras clave. Esta aproximación es fundamental para entender consultas formuladas de manera diferente pero con la misma intención.

**Ejemplo de similitud semántica**:
```
"¿Quién debe presentar el modelo 036?"
    ↓ embedding
[0.23, -0.45, 0.67, ..., 0.12]  # Vector 768D

"Obligados a declarar el alta censal"
    ↓ embedding  
[0.25, -0.43, 0.69, ..., 0.10]  # Alta similitud coseno (~0.95)
```

> ChromaDB es una base de datos vectorial de código abierto para almacenar y buscar embeddings


In [ ]:
# Importamos las librerías necesarias para ChromaDB
import chromadb
from chromadb.config import Settings
from chromadb.utils import embedding_functions

In [ ]:
# Configuración de ChromaDB
# Creamos un cliente persistente que guardará los datos en disco en la carpeta especificada
persist_directory = "./chroma_db"
chroma_client = chromadb.PersistentClient(path=persist_directory, settings=Settings(allow_reset=True))

2.1 Definimos funciones de embedding para vectorizar nuestros documentos

> Las funciones de embedding convierten texto en vectores numéricos que capturan el significado semántico



In [ ]:
# Función de embedding usando modelo de  modelo local gratuito
# sentence_transformers_ef = embedding_functions.SentenceTransformerEmbeddingFunction(model_name="all-MiniLM-L6-v2")
# Función de embedding usando modelo de Google (requiere API key)
google_ef = embedding_functions.GoogleGenerativeAiEmbeddingFunction(
    api_key=g_api_key,
    model_name="models/gemini-embedding-exp-03-07",
)

En ChromaDB, una colección es una especie de contenedor donde se guardan documentos vectorizados (es decir, representados como embeddings).

Es como una tabla en una base de datos relacional, donde cada "fila" es un documento con su embedding y metadatos.


In [ ]:
#Busca una colección llamada "documentos" en tu instancia de ChromaDB.
#Si ya existe, la recupera.
#Si no existe, la crea.
collection = chroma_client.get_or_create_collection(
    name="documentos_modelo36",
    embedding_function=sentence_transformers_ef
)

Agregamos los fragmentos de documentos a ChromaDB. Esto toma todos los fragmentos de texto del documento "modelo 36", los convierte en vectores, y los guarda en la colección m_36_collection en ChromaDB para futuras búsquedas semánticas.

In [ ]:
# Definimos una función para agregar múltiples documentos a ChromaDB de forma eficiente
# Esta función genera IDs únicos y metadatos básicos para cada documento
def agregar_documentos_a_chromadb(documents, collection):
    ids = [f"doc_{idx}" for idx in range(len(documents))]
    metadatas = [{"source": f"documento_{idx}"} for idx in range(len(documents))]

    print(f"Agregando {len(documents)} documentos de una sola vez.")
    collection.add(
        ids=ids,
        documents=documents,
        metadatas=metadatas
    )


In [ ]:
# Agregamos todos los fragmentos de texto del modelo 36 a la nueva colección
# Utilizamos la misma función de antes para mantener la consistencia
agregar_documentos_a_chromadb(modelo_36_raw_docs, collection)

Agregando 137 documentos de una sola vez.


# 3. Configurar el LLM y construir el sistema RAG

In [ ]:
import json

3.1 Búsqueda semántica (RAG Retrieval)

 El sistema utiliza similitud coseno para identificar los fragmentos más relevantes. Esta métrica mide el ángulo entre vectores, siendo invariante a la magnitud y efectiva para comparaciones en espacios de alta dimensionalidad.

```
similitud_coseno = (A · B) / (||A|| × ||B||)

Valores: 1.0 (idénticos) → 0.0 (ortogonales) → -1.0 (opuestos)
```

 Cuando un usuario hace una consulta, esta función busca los top_k fragmentos más relevantes en la colección collection usando similitud coseno.

In [ ]:
# Definimos un valor predeterminado más alto para top_k ya que el documento es extenso
DEFAULT_TOP_K = 6
# Función para buscar en la colección específica del modelo 36
def buscar_en_chromadb_m36(query, top_k=DEFAULT_TOP_K):
    """
    Realiza una búsqueda en la colección del modelo 36.

    Args:
        query: Texto de la consulta a buscar
        top_k: Número de resultados a devolver (por defecto 6 para este documento extenso)

    Returns:
        Resultados de la búsqueda en formato raw de ChromaDB
    """
    results = collection.query(
        query_texts=[query],
        n_results=top_k
    )
    return results

Herramientas y ejecución automática

> Esto permite mapear el nombre de la herramienta que el LLM puede usar a la función Python real que ejecuta la búsqueda.



In [ ]:
# Definición de la herramienta específica para buscar en la colección del modelo 36
tools_m36 = [
    {
        "type": "function",
        "function": {
            "name": "buscar_en_chromadb_m36",
            "description": "Busca información sobre el contenido del modelo 36, alta censal, de la agencia tributaria.",
            "parameters": {
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "description": "Consulta a realizar en la base de datos."
                    },
                    "top_k": {
                        "type": "integer",
                        "description": "Número máximo de resultados a devolver.",
                        "default": DEFAULT_TOP_K
                    }
                },
                "required": ["query"]
            },
        }
    }
]

# Diccionario que mapea el nombre de la función a la implementación real
tools_dict_m36 = {
    "buscar_en_chromadb_m36": buscar_en_chromadb_m36
}

Definición de la función que se utilizará en el RAG para llamar al LLM pasándole las herramientas disponibles.

El **objetivo** es obtener la respuesta inicial del LLM.


> Aquí el LLM (Gemini) puede decidir si usar una herramienta. En este caso, se le da acceso explícito a la función buscar_en_chromadb_m36



In [ ]:
def obtener_respuesta_modelo36(messages, herramientas_disponibles, modo_herramienta="required"):
    """
    Obtiene una respuesta del LLM utilizando la colección del modelo 36.

    Args:
        messages: Lista de mensajes para el contexto de la conversación
        herramientas_disponibles: Lista de herramientas disponibles para el LLM
        modo_herramienta: Estrategia de selección de herramientas ('required', 'auto', 'none')

    Returns:
        Objeto de respuesta del LLM
    """
    return openai.chat.completions.create(
        model="gemini-2.0-flash",
        messages=messages,
        tools=herramientas_disponibles,
        tool_choice=modo_herramienta,
    )


Ejecuta la herramienta elegida por el LLM y devuelve texto legible con contexto


> Si Gemini decide usar la herramienta, se ejecuta la búsqueda, y se devuelve el resultado formateado con metadatos (como el título del documento).




In [ ]:
# Función para formatear los resultados de ChromaDB de forma legible
# Toma el resultado crudo de ChromaDB y devuelve un texto con formato que incluye metadatos
def extraer_texto_con_metadata(resultado_chromadb: dict):
    documentos = resultado_chromadb.get("documents", [[]])[0]
    metadatas = resultado_chromadb.get("metadatas", [[]])[0]

    texto_plano = ""
    for doc, metadata in zip(documentos, metadatas):
        titulo = metadata.get("source", "Documento sin título")
        texto_plano += f"{titulo}:\n{doc}\n\n"

    return texto_plano.strip()


Definición de la función que permite ejecutar la herramienta seleccionada (búsqueda en base de datos, se utilizará para agregar el contexto recuperado al historial de mensajes en el RAG).

> Busca información relevante con la respuesta del LLM inicial en la base de datos vectorial del modelo 36



In [ ]:
def ejecutar_herramienta_modelo36(tool_call, diccionario_herramientas):
    """
    Ejecuta la herramienta seleccionada por el LLM para la colección del modelo 36.

    Args:
        tool_call: Objeto que contiene información sobre la herramienta a ejecutar
        diccionario_herramientas: Mapeo entre nombres de herramientas y funciones reales

    Returns:
        Texto formateado con los resultados de la búsqueda
    """
    # Extraer información de la llamada a la herramienta
    nombre_funcion = tool_call.function.name
    argumentos_json = tool_call.function.arguments

    # Obtener la función real a ejecutar
    funcion_real = diccionario_herramientas[nombre_funcion]

    # Convertir argumentos de JSON a diccionario Python
    argumentos = json.loads(argumentos_json)

    # Ejecutar la función y formatear el resultado
    resultado_crudo = funcion_real(**argumentos)
    return extraer_texto_con_metadata(resultado_crudo)

RAG (Retrieval-Augmented Generation) es un enfoque donde un modelo de lenguaje no responde solo con su conocimiento preentrenado, sino que primero recupera información relevante (desde ChromaDB en este caso) y usa eso como contexto para generar una respuesta.

Función **sistema_rag_modelo36()**. Esta función ejecuta un flujo completo de RAG para responder preguntas sobre el modelo 36.


1.   Usuario hace una pregunta
2.   LLM recibe el mensaje, y puede decidir usar una herramienta.

Si la usa:

*   Se ejecuta buscar_en_chromadb_m36
*   Se obtiene el texto relevante del documento
-   Se construye un nuevo mensaje con ese contexto
-   Se llama de nuevo al LLM (recursión) con el contexto ahora incluido

Si no usa herramientas, simplemente devuelve la respuesta generada.


## Pipeline RAG End-to-End

El sistema implementa un flujo recursivo inteligente donde el LLM decide autónomamente cuándo necesita información adicional. Esta arquitectura basada en herramientas (function calling) permite al modelo razonar sobre sus propias limitaciones de conocimiento.

#### Flujo de Ejecución:
1. **Input del usuario** → Query procesada
2. **LLM analiza** → Decide si requiere búsqueda de contexto
3. **Si necesita información**:
   - Ejecuta búsqueda semántica en ChromaDB
   - Recupera top-k fragmentos relevantes
   - Construye prompt enriquecido con contexto
4. **Generación final** → Respuesta fundamentada en documentación oficial

In [ ]:
def sistema_rag_modelo36(consulta_usuario="", historial_mensajes=None, modo_herramienta="required"):
    """
    Sistema RAG especializado para consultas sobre el modelo 36.

    Flujo completo del proceso RAG para el documento del modelo 36:
    1. Recibe la consulta del usuario sobre el modelo 36
    2. Busca información relevante en la base de datos vectorial del modelo 36
    3. Proporciona esa información como contexto al LLM
    4. Genera una respuesta final basada en ese contexto

    Args:
        consulta_usuario: Pregunta o instrucción del usuario sobre el modelo 36
        historial_mensajes: Lista de mensajes previos para mantener contexto
        modo_herramienta: Estrategia para el uso de herramientas por el LLM

    Returns:
        Diccionario con la respuesta final
    """
    # Inicializar historial de mensajes si no existe
    if historial_mensajes is None:
        historial_mensajes = []

    # Agregar la consulta del usuario al historial si existe
    if consulta_usuario:
        historial_mensajes.append({"role": "user", "content": consulta_usuario})

    # Paso 1: Obtener respuesta inicial del LLM
    # Llama al modelo (Gemini) con el historial y le dice: "puedes usar estas herramientas si quieres"
    # El modelo puede decidir si necesita buscar contexto antes de responde
    respuesta = obtener_respuesta_modelo36(
        messages=historial_mensajes,
        herramientas_disponibles=tools_m36,
        modo_herramienta=modo_herramienta
    )

    # Se comprueba si el LLM decidió usar una herramienta
    if respuesta.choices[0].message.tool_calls:
        # Paso 2: Ejecutar la herramienta seleccionada (búsqueda en base de datos)
        tool_call = respuesta.choices[0].message.tool_calls[0]
        contexto_recuperado = ejecutar_herramienta_modelo36(tool_call, tools_dict_m36)

        # Paso 3: Agregar el contexto recuperado al historial de mensajes
        mensaje_contexto = (
            "Contexto para responder:\n```" +
            contexto_recuperado +
            "\n```\nEn base a este contexto recuperado responde a la pregunta del usuario."
        )
        historial_mensajes.append({"role": "assistant", "content": mensaje_contexto})

        # Paso 4: Llamada recursiva para generar respuesta final basada en el contexto
        # Usamos 'auto' en lugar de 'none' para dar flexibilidad al LLM en la segunda etapa
        # Se hace una segunda llamada al modelo, ahora con el contexto, para que dé la respuesta final.
        return sistema_rag_modelo36(historial_mensajes=historial_mensajes, modo_herramienta="auto")
    else:
        # El LLM ya generó la respuesta final
        return {"respuesta": respuesta}

# 4. Generar respuestas

In [ ]:
# 4. Generar respuestas y actualizar evaluation_data directamente
for item in evaluation_data:
    user_input = item["user_input"]  # o "question" si es el nombre de la clave original
    resultado = sistema_rag_modelo36(consulta_usuario=user_input)

    # Extraer contenido de la respuesta
    respuesta_generada = resultado.get("respuesta").choices[0].message.content

    # Agregar la respuesta generada a la entrada original
    item["response"] = respuesta_generada

In [ ]:
evaluation_data

[{'id': 0,
  'user_input': '¿Quiénes están obligados a presentar la declaración censal de alta (Modelo 036)?',
  'reference': 'Están obligados a presentar la declaración de alta los empresarios o profesionales que vayan a comenzar actividades económicas en territorio español, quienes sin ser empresarios/profesionales abonen rentas sujetas a retención o realicen adquisiciones intracomunitarias de bienes sujetas al IVA, los no residentes que operen en España mediante establecimiento permanente o satisfagan rentas sujetas a retención, las entidades en régimen de atribución de rentas extranjeras con presencia en España, los socios/herederos/comuneros/partícipes de entidades en atribución de rentas con obligaciones tributarias derivadas, y los no establecidos en territorio de aplicación del IVA que sean sujetos pasivos del mismo (salvo exoneración).',
  'response': ' El modelo 036 es una declaración censal de alta, baja o modificación. No puedo determinar quiénes están obligados a presentar

# 5. Evaluar

Utilizar la función dada en el enunciado **evaluar_lista_respuestas()** que permite evaluar un conjunto de respuestas (una lista completa) y calcula:

- La precisión media del conjunto.
- Las precisiones individuales por cada ítem.

📤 Devuelve:

- Una tupla con:La precisión media (float)
- Un diccionario {id: puntuación_individual}.

🧠 Útil para:

- Evaluar el rendimiento global del sistema RAG.
- Generar informes de evaluación automáticos.

In [ ]:
puntuacion_media, puntuaciones_individuales = evaluar_lista_respuestas(evaluation_data)

print(f"Precisión media del sistema RAG: {puntuacion_media:.2f}")
print("\nResultados individuales:")
for id_, score in puntuaciones_individuales.items():
    print(f"Pregunta {id_}: {score:.2f}")

Precisión media del sistema RAG: 0.71

Resultados individuales:
Pregunta 0: 0.00
Pregunta 1: 0.75
Pregunta 2: 1.00
Pregunta 3: 0.75
Pregunta 4: 1.00
Pregunta 5: 0.75


## 📊 Resultados

### Evaluación en Dataset de Prueba

| ID | Pregunta | Precisión |
|----|----------|-----------|
| 0 | ¿Quiénes están obligados a presentar la declaración censal de alta? | 0.00 |
| 1 | ¿Para qué sirve la declaración censal de modificación? | 0.75 |
| 2 | ¿En qué plazo general se debe presentar la declaración de modificación? | 1.00 |
| 3 | ¿Dónde se puede presentar el modelo 036 en formato impreso? | 0.75 |
| 4 | ¿Cuándo es obligatorio presentar el modelo 036 por vía telemática? | 1.00 |
| 5 | ¿Qué plazo tienen los herederos para comunicar la modificación de titularidad? | 0.75 |

**Precisión Media del Sistema: 0.71 (71%)**

### Análisis de Resultados

✅ **Fortalezas**:
- Alta precisión en preguntas sobre plazos y procedimientos específicos
- Buena recuperación de contexto relevante
- Respuestas fundamentadas en el documento oficial

⚠️ **Áreas de Mejora**:
- Pregunta 0 requiere optimización del chunking (demasiado compleja)
- Posible mejora con ajuste del parámetro `top_k`
- Experimentar con diferentes estrategias de chunking
